docs/dependency-injection-liftweb-scala.adoc

```diff
@@ -0,0 +1,382 @@
+early.
+
+== Lift DI and various dependency scopes (e.g. from Spring)
+
+Lift's DI allows all the
+https://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-factory-scopes[four most commonly used scopes] that
+are possible with spring:
+https://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-factory-scopes-singleton[singleton],
+https://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-factory-scopes-prototype[prototype], https://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-factory-scopes-request[request],
+and https://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-factory-scopes-session[session]. The others can be achieved
+relatively easily if you want.
+
+== Getting started
+
+Lift's DI is based on two main traits https://liftweb.net/api/26/api/index.html#net.liftweb.util.Injector[`Injector`]
+and https://liftweb.net/api/26/api/index.html#net.liftweb.util.Maker[`Maker`]. However, most of the time, you won't need to
+deal with them directly. The elements that you would use are:
+
+. https://liftweb.net/api/26/api/index.html#net.liftweb.http.Factory[`Factory`] trait
+. https://github.com/lift/framework/blob/5033c8798d4444f81996199c10ea330770e47fbc/web/webkit/src/main/scala/net/liftweb/http/Factory.scala#L37[`FactoryMaker`], which is an abstract class inside
+ inside the [`Factory`] trait.
+. https://liftweb.net/api/26/api/index.html#net.liftweb.util.SimpleInjector$Inject[`Inject`], an abstract class within the
+ `SimpleInjector` trait.
+
+`FactoryMaker` and `Injector` serve the same purpose, with the former
+having more features. The difference is important. Notes on this are
+at the end. I'll continue with `FactoryMaker` for the sake of
+examples, and then note the differences between the two at the end.
+
+To start with, here's an example of how these are supposed to be
+used:
+
+[source,scala]
+----
+object DependencyFactory extends Factory {
+  private val seq = new AtomicLong(0)
+  
+  object emailService extends FactoryMaker(emailServiceImpl)
+  object sequenceNumberService extends FactoryMaker(seq.incrementAndGet _)
+  
+  private def emailServiceImpl: EmailService = Props.mode match {
+    case Props.RunModes.Production => new RealEmailService
+    // A stub for local development
+    case _ => new TestEmailService
+  }
+}
+----
+
+This defines two managed dependencies. In the case of the
+`emailService`, the dependency changes based on whether we are running
+on production/staging mode or some other mode (e.g. development). The
+`emailService` is a singleton dependency. That is, everyone
+who injects this dependency will get the same instance throughout the
+app. 
+
+Note that `FactoryMaker` constructor takes a stock scala function that
+returns an instance of the needed type. A http://www.scala-lang.org/api/current/scala/Function0.html[`Function0`] to
```

Comment:
Should be `needed type---a Function0, to be precise`; otherwise, we have a fragment.